File: Role.m
//
//  Role.m
//  mage-ios-sdk
//
//  Created by William Newman on 4/13/16.
//  Copyright © 2016 National Geospatial-Intelligence Agency. All rights reserved.
//

#import "Role.h"
#import "MageServer.h"
#import "HttpManager.h"

@implementation Role

+ (Role *) insertRoleForJson: (NSDictionary *) json inManagedObjectContext:(NSManagedObjectContext *) context {
    Role *role = [Role MR_createEntityInContext:context];
    [role updateRoleForJson:json];
    
    return role;
}

- (void) updateRoleForJson: (NSDictionary *) json {
    [self setRemoteId:[json objectForKey:@"id"]];
    [self setPermissions:[json objectForKey:@"permissions"]];
}

+ (NSURLSessionDataTask *) operationToFetchRolesWithSuccess:(void (^ _Nullable)()) success
                                           failure:(void (^ _Nullable)(NSError *error)) failure {
    NSString *url = [NSString stringWithFormat:@"%@/%@", [MageServer baseURL], @"api/roles"];
    
    NSLog(@"Trying to fetch users from server %@", url);
    
    HttpManager *http = [HttpManager singleton];
    NSURLSessionDataTask *task = [http.manager GET:url parameters:nil progress:nil success:^(NSURLSessionTask *task, id roles) {
        [MagicalRecord saveWithBlock:^(NSManagedObjectContext *localContext) {
            // Get the user ids to query
            NSMutableArray *roleIds = [[NSMutableArray alloc] init];
            for (NSDictionary *roleJson in roles) {
                [roleIds addObject:[roleJson objectForKey:@"id"]];
            }
            
            NSArray *rolesMatchingIds = [Role MR_findAllWithPredicate:[NSPredicate predicateWithFormat:@"(remoteId IN %@)", roleIds] inContext:localContext];
            NSMutableDictionary *roleIdMap = [[NSMutableDictionary alloc] init];
            for (Role* role in rolesMatchingIds) {
                [roleIdMap setObject:role forKey:role.remoteId];
            }
            
            for (NSDictionary *roleJson in roles) {
                // pull from query map
                NSString *roleId = [roleJson objectForKey:@"id"];
                Role *role = [roleIdMap objectForKey:roleId];
                if (role == nil) {
                    // not in core data yet need to create a new managed object
                    NSLog(@"Inserting new role into database");
                    [Role insertRoleForJson:roleJson inManagedObjectContext:localContext];
                } else {
                    // already exists in core data, lets update the object we have
                    NSLog(@"Updating role in the database");
                    [role updateRoleForJson:roleJson];
                }
            }
        } completion:^(BOOL contextDidSave, NSError *error) {
            if (error) {
                if (failure) {
                    failure(error);
                }
            } else if (success) {
                success();
            }
        }];

    } failure:^(NSURLSessionTask *operation, NSError *error) {
        if (failure) {
            failure(error);
        }
    }];
    
    return task;
}
@end

File: Mage/CoreData/Role.swift
//
//  Role.m
//  mage-ios-sdk
//
//  Created by William Newman on 4/13/16.
//  Copyright © 2016 National Geospatial-Intelligence Agency. All rights reserved.
//

import Foundation
import CoreData

@objc public class Role: NSManagedObject {
    
    @discardableResult
    @objc public static func insert(json: [AnyHashable : Any], context: NSManagedObjectContext) -> Role? {
        let role = Role.mr_createEntity(in: context);
        role?.update(json: json, context: context);
        return role;
    }
    
    @objc public func update(json: [AnyHashable : Any], context: NSManagedObjectContext) {
        self.remoteId = json[RoleKey.id.key] as? String
        self.permissions = json[RoleKey.permissions.key] as? [String]
    }
    
    @objc public static func operationToFetchRoles(success: ((URLSessionDataTask,Any?) -> Void)?, failure: ((URLSessionDataTask?, Error) -> Void)?) -> URLSessionDataTask? {
        guard let baseURL = MageServer.baseURL() else {
            return nil
        }
        let url = "\(baseURL.absoluteURL)/api/roles";
        let manager = MageSessionManager.shared();
        let methodStart = Date()
        NSLog("TIMING Fetching Roles @ \(methodStart)")
        let task = manager?.get_TASK(url, parameters: nil, progress: nil, success: { task, responseObject in
            NSLog("TIMING Fetched Roles. Elapsed: \(methodStart.timeIntervalSinceNow) seconds")
            if let responseData = responseObject as? Data {
                if responseData.count == 0 {
                    print("Roles are empty");
                    success?(task, nil);
                    return;
                }
            }
            
            guard let roles = responseObject as? [[AnyHashable : Any]] else {
                success?(task, nil);
                return;
            }
            let saveStart = Date()
            NSLog("TIMING Saving Roles @ \(saveStart)")
            MagicalRecord.save { localContext in

                // Get the role ids to query
                var roleIds: [String] = [];
                for roleJson in roles {
                    if let roleId = roleJson[RoleKey.id.key] as? String {
                        roleIds.append(roleId);
                    }
                }

                let rolesMatchingIDs: [Role] = Role.mr_findAll(with: NSPredicate(format: "(\(RoleKey.remoteId.key) IN %@)", roleIds), in: localContext) as? [Role] ?? [];
                var roleIdMap: [String : Role] = [:];
                for role in rolesMatchingIDs {
                    if let remoteId = role.remoteId {
                        roleIdMap[remoteId] = role
                    }
                }
                
                for roleJson in roles {
                    // pull from query map
                    guard let roleId = roleJson[RoleKey.id.key] as? String else {
                        continue;
                    }
                    if let role = roleIdMap[roleId] {
                        // already exists in core data, lets update the object we have
                        print("Updating role in the database \(role.remoteId ?? "")");
                        role.update(json: roleJson, context: localContext);

                    } else {
                        // not in core data yet need to create a new managed object
                        print("Inserting new role into database");
                        Role.insert(json: roleJson, context: localContext)
                    }
                }
            } completion: { contextDidSave, error in
                NSLog("TIMING inserted roles. Elapsed: \(saveStart.timeIntervalSinceNow) seconds")

                if let error = error {
                    if let failure = failure {
                        failure(task, error);
                    }
                } else if let success = success {
                    success(task, nil);
                }
            }
        }, failure: { task, error in
            if let failure = failure {
                failure(task, error);
            }
        });
        return task;
    }
}
